File: Implementações/InsertionSort.c
#include <stdio.h>
#include <time.h>
#include <stdlib.h>

// Implementação do Insertion Sort
void insertionSort(int arr[], int n)
{
  for (int i = 1; i < n; i++)
  {
    int key = arr[i];
    int j = i - 1;

    // Move elementos maiores que a "key" uma posição à frente
    while (j >= 0 && arr[j] > key)
    {
      arr[j + 1] = arr[j];
      j = j - 1;
    }
    arr[j + 1] = key;
  }
}

// Função para gerar dados aleatórios
void ArrayAleatorio(int arr[], int n)
{
  for (int i = 0; i < n; i++)
  {
    arr[i] = rand() % 10000;
  }
}

// Função para exibir o array
void printArray(int arr[], int n)
{
  for (int i = 0; i < n; i++)
  {
    printf("%d ", arr[i]);
  }
  printf("\n");
}

int main()
{
  int n = 100000; // Tamanho do array (você pode alterar para testar diferentes tamanhos)
  int arr[n];

  // Gerar array aleatório
  ArrayAleatorio(arr, n);

  printf("Array original:\n");
  printArray(arr, n);

  // Medir o tempo de execução
  clock_t inicio, fim;
  inicio = clock();
  insertionSort(arr, n);
  fim = clock();

  printf("Array ordenado:\n");
  printArray(arr, n);

  double tempo_gasto = ((double)(fim - inicio)) / CLOCKS_PER_SEC;
  printf("Tempo de execucao: %f segundos\n", tempo_gasto);

  return 0;
}

File: ImplementaçõesPython/InsertionSort.py
import time
import random

# Implementação do Insertion Sort
def insertion_sort(arr):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key

# Função para gerar dados aleatórios
def gerar_array_aleatorio(n, max_value=10000):
    return [random.randint(0, max_value) for _ in range(n)]

# Main
if __name__ == "__main__":
    n = 10000  # Tamanho do array (mude para testar diferentes tamanhos)
    arr = gerar_array_aleatorio(n)

    print("Array original:")
    print(arr)

    # Medir o tempo de execução
    inicio_tempo = time.time()
    insertion_sort(arr)
    fim_tempo = time.time()

    print("Array ordenado:")
    print(arr)

    print(f"Tempo de execução: {fim_tempo - inicio_tempo:.6f} segundos")
